posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,324 @@
+    input.df <- data.frame(input.mat)
+    input.csv <- tempfile()
+    fwrite(input.df, "data.csv")
+    pd <- import("pandas")
+    input_df_pd <- pd$DataFrame(input.df) 
+  },
+  seconds.limit = seconds.limit,
+  "data.table::fread" = {
+    data.table::fread("data.csv", showProgress = FALSE) 
+  },
+  "pandas.read_csv " = {
+    pd <- import("pandas")
+    reticulate::py_run_string("import pandas as pd")
+    reticulate::py_run_string("pd.read_csv(file_path)")  
+  }
+)
+```
+
+```{r,warning=FALSE,message=FALSE}
+refs.read.vary.cols <- atime::references_best(atime.read)
+```
+
+```{r,warning=FALSE,message=FALSE}
+pred.read.vary.cols <- predict(refs.read.vary.cols)
+plot(pred.read.vary.cols)
+```
+
+```{r,warning=FALSE,message=FALSE}
+gg.read.pd <- plot(pred.read.vary.cols)+
+  theme(text=element_text(size=15))+
+  ggtitle(sprintf("Read real numbers to CSV, with pandas in Python \nand data.table in R, %d x N", n.rows))+
+  scale_x_log10("N = number of columns to write")+
+  scale_y_log10("Computation time (seconds)
+median line, min/max band
+over 10 timings")+
+  facet_null()+
+  scale_fill_manual(values=read.colors)+
+  scale_color_manual(values=read.colors)
+```
+
+```{r,warning=FALSE,message=FALSE}
+plot(gg.read.pd)
+```
+
+When it comes to reading data, data.table in R also demonstrates its superiority. It provides fast and efficient methods for importing and reading various file formats, including CSV. The fread() function in data.table is known for its speed and memory efficiency, making it an optimal choice for handling large datasets.
+
+# Example 3. Reshape performance comparison.
+
+Data reshaping means changing the shape of the data, to get it into a more appropriate format, for learning/plotting/etc. Here we consider wide to long  and long to wide reshape, which means we start with a wide table (many columns) and end up with a long table (fewer columns) and vice versa. 
+
+## A. wide to long reshape.
+
+In data.table, the data.table::melt() function is used to convert data from a wide format to a long format, while in Pandas, the pandas::melt() function is used to convert data from a wide format to a long format.
+
+## Comparison code
+## data.table::melt() is faster
+
+```{r,warning=FALSE,message=FALSE,results='hide'}
+ml.colors <- c(
+  "data.table::melt"="#D6604D",
+  "pd.melt" = "#BF812D"
+  )
+n.folds <- 10
+n.rows <- 100
+seconds.limit <- 10
+ml.reshape.atime <- atime::atime(
+  N=as.integer(10^seq(2, 15, by=0.2)),
+  setup={
+    df <- data.frame(
+      id = rep(1:N, each = 2),
+      category = rep(c("A", "B"), N),
+      value = rnorm(2 * N)
+      )
+    },
+  seconds.limit= seconds.limit,
+  "data.table::melt" = {
+    data.table::melt(data.table(df), id.vars = c("id",  "category"),variable.names="variable", value.name = "value")
+  },
+  "pd.melt" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
+    pd$melt(py_df, id_vars = c("id", "category"), value_name = "score")  
```

Comment:
why is value_name="score" here but value.name="value" above?

Reply:
addressed